Add support for PlantsIO Ivy smart planter

thewade opened the pull request:
This PR adds the PlantsIO "Ivy" Smart Planter.

Info on the device:
https://www.indiegogo.com/projects/plantsio-ivy-smart-flowerpots-endless-fun#/
https://smartplantivy.com/

Available on Amazon and AliExpress. Lots of drop shippers using various brand names.

Datapoints documented at https://gist.github.com/thewade/ef9f6014f13932bd2e77d43331e2027d

This device has a ton of sensors and config options. I haven't got them all the config done, and some enum data needs to be reverse engineered to be mapped.

Some settings like brightness should be restricted when auto brightness is on, but I wasn't able to figure out if it was possible using constraints and conditions.

I expect another PR in the future to fill in the missing details as well as constraints and conditions if they are possible, but I need a break for now.

make-all commented on custom_components/tuya_local/devices/plantsio_ivy_smart_planter.yaml:
```diff
@@ -0,0 +1,431 @@
+    name: PlantsIO Ivy Smart Planter
+primary_entity:
+  entity: sensor
+  name: Water level
```
`class: volume_storage`

make-all commented on custom_components/tuya_local/devices/plantsio_ivy_smart_planter.yaml:
```diff
@@ -0,0 +1,431 @@
+secondary_entities:
+  # Sensor entities
+  - entity: sensor
+    name: Ambient light
```
Suggest to remove this name, so that pre-defined translations can be used.

thewade replied:
When I do that all my sensors have no names:

![image](https://github.com/make-all/tuya-local/assets/744796/2cb5f4ad-5fb9-4a38-a8c7-bac438479897)

And I end up getting entity IDs like:

sensor.smart_planter
sensor.smart_planter_2
sensor.smart_planter_3

Am I doing something wrong?

This is my current code for the illuminance sensor:

```
  - entity: sensor
    class: illuminance
    dps:
      - id: 103
        type: integer
        name: sensor
        unit: lx
        class: measurement
```

make-all replied:
This is maybe a recently introduced bug due to a side effect of the fix for #1856.  It may be better for me to revert that change.
Unfortunately this was difficult to detect, as existing entity names do not change once they are setup, only newly added ones are affected.

make-all commented on custom_components/tuya_local/devices/plantsio_ivy_smart_planter.yaml:
```diff
@@ -0,0 +1,431 @@
+        unit: lx
+        class: measurement
+  - entity: sensor
+    name: Temperature
```
Definitely remove the names that are the same as class

make-all commented on custom_components/tuya_local/devices/plantsio_ivy_smart_planter.yaml:
```diff
@@ -0,0 +1,431 @@
+          - value: "Unknown state"
+  - entity: sensor
+    name: Humidity status
+    class: enum
```
Sensors with only two states could maybe be made into binary_sensor instead, especially if there is a relevant class defined for binary_sensor.

thewade replied:
I would expect there to be multiple states from the "AI" derived sensors. 

e.g. Good, Dry, Too humid, Too dry

However I have no been able to determine all the values in my environment, and as I see with Light status enum the values are not in order so I don't want to guess.

The developers seems to had expressed interested in home assistant status, so I have emailed them to see if they are willing to share the mapping.

make-all commented on custom_components/tuya_local/devices/plantsio_ivy_smart_planter.yaml:
```diff
@@ -0,0 +1,431 @@
+        name: sensor
+        unit: "%"
+  - entity: sensor
+    name: Charge state
```
This can definitely be a binary_sensor (`class: battery_charging`), since the original value is boolean even

make-all commented on custom_components/tuya_local/devices/plantsio_ivy_smart_planter.yaml:
```diff
@@ -0,0 +1,431 @@
+        type: boolean
+        name: switch
+  - entity: number
+    name: Brightness
```
Is this a light?

thewade replied:
This is the display brightness.

thewade commented:
I have added a commit with the following changes:

- fix most enums to use integer instead of bitfield (some states were not mapping until I fixed it)
- use volume_storage class for water level
- add absorbing state to water status
- add initializing states from missing enums
- use binary_sensor entity and battery_charging class for charging status
- convert plant in pot to binary_sensor
- add not selected enum for plant type
- add config options for weather and weather location

I haven't made the change regarding the names of known class types due the issue I am seeing and I will wait for your feedback.

thewade commented:
> * convert plant in pot to binary_sensor

There seems to be some issue with this change as I now have an error in the logs:

```
This error originated from a custom integration.

Logger: custom_components.tuya_local.sensor
Source: custom_components/tuya_local/sensor.py:59
integration: Tuya Local (documentation, issues)
First occurred: 10:52:57 AM (2155 occurrences)
Last logged: 5:24:13 PM

plantios_ivy_smart_planter.yaml/Plant in pot: Unrecognized sensor device class of problem ignored
```

The entry:
```
  - entity: sensor
    name: Plant in pot
    class: problem
    category: diagnostic
    dps:
      - id: 102
        type: boolean
        name: sensor
```
